### ISSUE TITLE
Custom errors not picked up by Hardhat Network when using external Hardhat artifact

### ISSUE THREAD

**Issue by paulrberg** [NomicFoundation/hardhat#1618]
## Description

If I import an artifact from an external package and use Waffle's "deployContract" to deploy it to my local network, and that contract reverts with a custom error, the custom error won't be picked up by Hardhat when running tests.

## How to Reproduce

In the terminal:

- git clone git@github.com:paulrberg/hardhat-template.git
- cd hardhat-template
- yarn
- yarn add [@hifi/protocol](https://unpkg.com/browse/@hifi/protocol@1.1.1)

Then write a test that imports the [FintrollerV1](https://github.com/hifi-finance/hifi-protocol/tree/9aed645ef6c867790d71886a4f433ec3134153aa/contracts/core/fintroller) contract and calls the [setMaxBonds](https://github.com/hifi-finance/hifi-protocol/blob/9aed645ef6c867790d71886a4f433ec3134153aa/contracts/core/fintroller/FintrollerV1.sol#L264-L268) function. The execution should revert with the [OwnableUpgradeable__NotOwner](https://github.com/hifi-finance/hifi-protocol/blob/9aed645ef6c867790d71886a4f433ec3134153aa/contracts/access/OwnableUpgradeable.sol#L9) custom error, but it doesn't. It reverts with the following error:

> Error: Transaction reverted without a reason string

## Environment

- hardhat@2.4.1
- @hifi/protocol@1.1.1

**Comment by fvictorio** [comment]
(Pasting my answer from discord here)

It's a known limitation: if a contract is not part of your project, you don't have the information needed to parse the custom error.

I don't think there's an easy way to fix that. We want to, at some point, add the capability of using "external" information (e.g., from etherscan), but it's not something we are working on right now

The easy fix is to add that contract to your project, for example using [hardhat-dependency-compiler](https://hardhat.org/plugins/hardhat-dependency-compiler.html)

**Comment by paulrberg** [comment]
Oooh, right. I remember reading about this limitation in the [release spec](https://github.com/nomiclabs/hardhat/releases/tag/hardhat-core-v2.4.0):

> Keep in mind that Hardhat can only recognize custom errors that are defined within your project. If you use mainnet forking and call some external contract that reverts with a custom error, Hardhat won't be able to parse it, and it will show an unrecognized error instead.

But I forgot about it.

**Comment by paulrberg** [comment]
Side note: what if Hardhat included the error selector in the logged error? That way, those of us who know how to handle selectors could use them. It happens that ethers.js has a [getSighash](https://docs.ethers.io/v5/api/utils/abi/interface/#Interface--selectors) method for grabbing the 4byte selector.

**Comment by alcuadrado** [comment]
Is this delegatecalling into an external contract? Or is the target one also yours?

**Comment by paulrberg** [comment]
The target contract is also mine. It is defined in the same repo and I can see an artifacts generated for it. See this:

https://github.com/paulrberg/prb-proxy/tree/6f2e4c2da37c3289d85218692d5a0a7f3a0ebb8f/contracts/test

> I think this is going to be increasingly more and more important as the custom errors start to get used on mainnet.

Yes! Especially true after OpenZeppelin [adds](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2839) support for them.

**Comment by fvictorio** [comment]
Opened #1875 based on this issue. I didn't edit this one because I'm not 100% sure that it's exactly the same problem.

**Comment by paulrberg** [comment]
I just found another case where the lack of support for "external" custom errors is problematic: when using the [hardhat_setCode](https://hardhat.org/hardhat-network/reference/#hardhat-setcode) JSON-RPC Method.

When I deploy a mock USDC contract the usual way, my tests pass because custom errors are picked up by Hardhat. But when I deploy via this code snippet:

```ts
```ts
export async function deployUsdc(deployer: Signer): Promise<GodModeErc20> {
  const godModeErc20Artifact: Artifact = await artifacts.readArtifact("GodModeErc20");

  // Deploy via CREATE first, to get a runtime bytecode with the constructor arguments.
  const godModeErc20: GodModeErc20 = <GodModeErc20>(
    await waffle.deployContract(deployer, godModeErc20Artifact, [USDC_NAME, USDC_SYMBOL, USDC_DECIMALS])
  );

  // Get the runtime bytecode.
  const runtimeBytecode: string = await ethers.provider.getCode(godModeErc20.address);

  // Deploy the USDC contract at address(1).
  await ethers.provider.send("hardhat_setCode", [USDC_ADDRESS, runtimeBytecode]);

  // Load the USDC contract at address(1);
  const usdc: GodModeErc20 = GodModeErc20__factory.connect(USDC_ADDRESS, deployer);

  return usdc;
}
```

Tests fail because of this error:

```
Error: Transaction reverted without a reason string
      at <UnrecognizedContract>.<unknown> (0x051e5b9fa1e3149e44dec9aa2c61af79f696afd8)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at runNextTicks (node:internal/process/task_queues:65:3)
      at listOnTimeout (node:internal/timers:526:9)
      at processTimers (node:internal/timers:500:7)
      at async EthModule._estimateGasAction (/Users/paulrberg/workspace/hifi/monorepo/node_modules/hardhat/src/internal/hardhat-network/provider/modules/eth.ts:425:7)
      at async HardhatNetworkProvider.request (/Users/paulrberg/workspace/hifi/monorepo/node_modules/hardhat/src/internal/hardhat-network/provider/provider.ts:108:18)
      at async EthersProviderWrapper.send (/Users/paulrberg/workspace/hifi/monorepo/node_modules/@nomiclabs/hardhat-ethers/src/internal/ethers-provider-wrapper.ts:13:20)
```

It would be nice if there was a way to "map" the contracts deployed with `setCode` so that custom errors are retained by the tracing engine.

**Comment by wei3erHase** [comment]
i understand the technical situation, but is there a way of reading those CustomErrors from the interface?

**Comment by wei3erHase** [comment]
Let me rephrase:
The interfaces already has the CustomErrors
We're fetching a contract with an interface and an address
When something happens in that contract, we receive a revert `without revert string` (instead of not being able to decode it)

Case scenario:
We're looking to make a deployment for developers without showing the source code (without validating in Etherscan) but if we provide only the interface (which has all the CustomErrors) the functionality does not work.

Workaround:
We've made a find and replace:
`revert CustomError()` >>> `revert('CustomError()')`

**Comment by PatrickAlphaC** [comment]
@paulrberg how did you end up testing for the custom errors in hardhat? Do you have an example?

**Comment by paulrberg** [comment]
@PatrickAlphaC I ended up importing the said contract which included the custom error I wanted to test for. I did that by [manually importing the contract in my code base](https://github.com/hifi-finance/hifi/blob/fb506210174c88eaa521241221e8c7084a2d1416/packages/flash-swap/contracts/test/Fintroller.sol), but if you'd like to not write a dummy contract you could instead use [hardhat-dependency-compiler](https://www.npmjs.com/package/hardhat-dependency-compiler).

At any rate, this would be much easier to handle if Hardhat managed somehow to decode custom errors from external contracts.

**Comment by fvictorio** [comment]
> At any rate, this would be much easier to handle if Hardhat managed somehow to decode custom errors from external contracts.

This would be great although really hard to do in the general case, because you only see a bunch of bytes and you need to guess the signature of the error.

Some things we could do are:

- Have a list of "well-known" custom errors that comes with Hardhat
- Fetch the code from etherscan, assuming it's verified

But if you are using some obscure contract that is not verified, I think it's just impossible to guess the error.

**Comment by paulrberg** [comment]
> Have a list of "well-known" custom errors that comes with Hardhat

This would be epic! Something like [4byte.directory](https://www.4byte.directory/) but for custom errors.

> Fetch the code from etherscan, assuming it's verified

Also cool.

> But if you are using some obscure contract that is not verified, I think it's just impossible to guess the error.

That's fine.

**Comment by fvictorio** [comment]
> Something like [4byte.directory](https://www.4byte.directory/) but for custom errors.

You probably know this already, but the same mechanism is used. That is, custom errors are encoded in the same way that function calls are encoded. So 4byte can include custom errors signatures too.

**Comment by paulrberg** [comment]
>  That is, custom errors are encoded in the same way that function calls are encoded

I actually didn't know, I didn't read the low-level spec. Fantastic.

**Comment by paulrberg** [comment]
@IAliceBobI custom errors don't exist in v0.6 of Solidity. They have been [introduced](https://blog.soliditylang.org/2021/04/21/custom-errors/) in v0.8.4.

There's really no good reason to not upgrade to the latest compiler version.

**Comment by IAliceBobI** [comment]
> @IAliceBobI custom errors don't exist in v0.6 of Solidity. They have been [introduced](https://blog.soliditylang.org/2021/04/21/custom-errors/) in v0.8.4.
> 
> There's really no good reason to not upgrade to the latest compiler version.

I want to upgrade, but I depend on many 0.6 Solidity audited libraries right now.
I can handle the error if I can get the raw error bytes.

**Comment by fvictorio** [comment]
@IAliceBobI maybe https://github.com/NomicFoundation/hardhat/pull/2553 can help you with that when it's released.

**Comment by IAliceBobI** [comment]
> @IAliceBobI maybe #2553 can help you with that when it's released.

I depend on this branch locally, and it works fine!
Thank you so much!

**Comment by kevincheng96** [comment]
> I've just bumped into a case where this happens even when the custom errors is defined locally in my project.
> 
> I have a custom error defined in contract B, which contract A interacts with via DELEGATECALL. When contract B reverts with that custom error, I am getting this error in the console when running my test:
> 
> > AssertionError: Expected transaction to be reverted with TargetError, but other exception was thrown: Error: VM Exception while processing transaction: reverted with an unrecognized custom error

@paulrberg Curious if you found a fix or workaround for the delegatecall case? I've also been running into this issue recently 🤔

**Comment by paulrberg** [comment]
Hi @kevincheng96, I can't remember the exact issue I described above, but I think I finally made it work at least in my testing environment. Take a look at the [execute.ts](https://github.com/paulrberg/prb-proxy/blob/54bfc578713bfe8b5df7e721ec79ccb65ecb78bd/test/contracts/prbProxy/effects/execute.test.ts) file in the [prb-proxy](https://github.com/paulrberg/prb-proxy) repository.

**Comment by kevincheng96** [comment]
> Hi @kevincheng96, I can't remember the exact issue I described above, but I think I finally made it work at least in my testing environment. Take a look at the [execute.ts](https://github.com/paulrberg/prb-proxy/blob/54bfc578713bfe8b5df7e721ec79ccb65ecb78bd/test/contracts/prbProxy/effects/execute.test.ts) file in the [prb-proxy](https://github.com/paulrberg/prb-proxy) repository.

Great, will take a look at it. Thanks for quick response!

**Comment by paulrberg** [comment]
Looks like @samczsun newly released signature database could be relevant here?

[sig.eth.samczsun.com](https://sig.eth.samczsun.com/)

**Comment by TsigelnikovNikita** [comment]
> @IAliceBobI maybe #2553 can help you with that when it's released.

Hello! In what hardhat version can I find these changes?

**Comment by fvictorio** [comment]
@TsigelnikovNikita I don't remember exactly which version of Hardhat added it, but it should be available in the latest versions.

**Comment by GeraldHost** [comment]
Any plans to make it so we can pass in selectors? 

I run into this problem with running in a fork:
- The contract I am calling has an artifact
- connect with typechain `Contract__factory.connect(addr, signer)` 
- expect call to error with custom error:
   - Instead get this error message `reverted with an unrecognized custom error`
   - When I inspect the stack trace I can see that the selector is the custom error selector

It would be good If we could just do something like:

`.to.be.revertedWith("0xe2e1f17f");`

**Comment by fvictorio** [comment]
> .to.be.revertedWith("0xe2e1f17f");

That's not a bad idea. The name of a custom error can't start with `0`, so there wouldn't be any ambiguity. Feel free to open a new issue with that feature request.

**Comment by fvictorio** [comment]
The syntax would be `.revertedWithCustomError("0x12345678")` though

**Comment by fvictorio** [comment]
I'm going to close this in favor of https://github.com/NomicFoundation/hardhat/issues/3486, so that we have a single issue tracking this need.
